osu.Game/Screens/Play/SongProgressGraph.cs

```diff
@@ -6,43 +6,207 @@
+
+                for (int x = 0; x < Strains.Count; x++)
+                {
+                    if (values[x]==0)
+                        Strains[x] = Strains.Max() * (-1);
+                    else
+                    {
+                        if (Strains[x]<0.01)
+                        {
+                            Strains[x]=1;
+                        }
+                    }
+                }
+                Update();
+            }
+        }
+
+        private double strainStep;
+
+        public double StrainStep
+        {
+            get { return strainStep; }
+            set
+            {
+                strainStep = value;
+            }
+        }
+
+        private IClock audioClock;
+
+        public IClock AudioClock
+        {
+            get { return audioClock; }
+            set { audioClock = value; }
+        }
+
+        private List<BreakPeriod> breaks;
+
+        public List<BreakPeriod> Breaks
+        {
+            get { return breaks; }
+            set
+            {
+                breaks = value;
+
+                var startOfLists = (objects.First().StartTime - objects.First().StartTime % (strainStep * audioClock.Rate))/(strainStep * audioClock.Rate);
+                if (strainStep==1)
+                {
+                    startOfLists = objects.First().StartTime;
+                }
+
+                var interval = strainStep;
+                if (interval==1)
+                {
+                    strainStep = 400;
+                }
+
+                for (int x = 0; x < strains.Count; x++)
+                {
+                    for (int i = 0; i < breaks.Count; i++)
+                    {
+                        if ( breaks[i].Duration >= interval / 2)
+                        {
+                            double startOfTouch = -1;
+                            if ( x * interval + startOfLists >= breaks[i].StartTime && x * interval + startOfLists < breaks[i].EndTime)
+                            {
+                                startOfTouch = x * interval + startOfLists;
+                            }
+                            else
+                            {
+                                if ( x * interval + startOfLists < breaks[i].StartTime)
+                                {
+                                    if (breaks[i].StartTime - (x * interval + startOfLists) < interval /2)
+                                    {
+                                        startOfTouch = breaks[i].StartTime;
+                                    }
+                                    else
+                                        continue;
+                                }
+                                else
+                                    continue;
+                            }
+
+                            double endOfTouch = -1;
+                            if ( (x+1) * interval + startOfLists >= breaks[i].StartTime && (x+1) * interval + startOfLists < breaks[i].EndTime)
```

Comment:
Simplify expression by storing `(x + 1) * interval + startOfLists` in a local constant